### PR TITLE
docs/design: add index acceleration

### PR DESCRIPTION
**Pull request by Benjamin2037**
### What problem does this PR solve?
Issue Number: ref #35983

Problem Summary:

### What is changed and how it works?
This is a pr to new design doc of add index acceleration.

### CODE REVIEW

File: docs/design/2022-06-06-Adding-Index-Accelerate.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+      
+### Limitations
+As the first version, we have some limitations. in the feture, we will keep improve the solution.
+1. This solution will be an expirement feature released in TiDB 6.2.
+2. Checkpoint currently not support, that means we may restart the backfill process in TiDB DDL Owner restart scenario.
+3. The lightning need cpu and memory to make backfill faster, so currently it may consume more CPU. We have limited memory to be over used and will continue implement a flexible resource control framework for DDL.
+
+### Enable/Disable
+1. Use tidb_fast_ddl to enable and disable new solution.  
+```TiDB> set global tidb_fast_ddl = on/off;```
```
Comment (morgo) [inline docs/design/2022-06-06-Adding-Index-Accelerate.md:111]:
```suggestion
```TiDB> set global tidb_enable_fast_ddl = on/off;```
```
This is existing the convention for boolean option flags.

File: docs/design/2022-06-06-Adding-Index-Accelerate.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+                                  | (reorgCtx.doneCh)
+                                  |
+                            worker master
+                                  ^ (waitTaskResults)
+                                  |
+                                  |
+                                  v (sendRangeTask)
+    +--------------------+---------+---------+------------------+-----------------+
+   |                    |                   |                  |                 |
+ backfillworker1     backfillworker2     backfillworker3     backfillworker4     ...
```
Comment (tangenta) [inline docs/design/2022-06-06-Adding-Index-Accelerate.md:28]: This is not well rendered.

File: docs/design/2022-06-06-Adding-Index-Acceleration.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+3. The index data is sorted when compacting multiple SST files into one, each sorted files will be ingested into TiKV by the existing interface provided.
+4. TiKV will finally combine the full copy of index data that was ingested with delta index part that was updated by user DML transaction during full copy preparation period. As the index at this time is almost completed its build process, so TiKV work in this step is not much.
+
+Since Lightning writes index data firstly into local storage instead of committing backfill transactions to TiKV. This removes conflict with user transactions and there is no transaction commit time for backfill also. The main path bottleneck of adding indexes is removed. 
+
+In addition to performance enhancement, the new solution could further lower the impact that DDL caused on DML. This point will be outstanding in the future implementation plan.
+
+## Uniqueness check
+For unique index / primary key, we should be able to detect if duplicate key is generated and reports error if so. In order to archive that, we could reuse the `DuplicateDetect` interface provided by the Lightning engine:
```
Comment (zimulala) [inline docs/design/2022-06-06-Adding-Index-Acceleration.md:59]:
```suggestion
For a unique index / primary key, we should be able to detect if a duplicate key is generated and reports an error if so. In order to archive that, we could reuse the `DuplicateDetect` interface provided by the Lightning engine:
```

**Review by tangenta — APPROVED** [review]

File: docs/design/2022-06-06-Adding-Index-Acceleration.md [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+- WriteReorganization - the index schema object is visible internally for the transactions, and index record can be added/removed/updated and the backfill workers are scheduled in this state to make sure index record will be 'backfilled' all for existing rows.
+- Public - the final state in which all work is done and the index is visible to the user.
+
+The backfill process scans all records from the table and generates an index key/value pair for each record, then fills the index data back into TiKV to form new index data. We found that backfilling data for index is a time-consuming task and it will be longer with user table data expanding sharply.
+
+There are some measures that have been done to enhance the performance of creating index statements. For example, 1) the parallel execution backfill tasks and 2) adjustable batch size for every backfill batch. Below figure 1 is used to show the current backfill implementation.
+
+![figure 1](./imgs/addIndexLit-1.png)
+
+From figure 1, the backfill task is firstly split into serials sub tasks, each sub tasks include one batch size record. Then, multi-workers are started up to execute backfill sub tasks parallely.
```
Comment (zimulala) [inline docs/design/2022-06-06-Adding-Index-Acceleration.md:22]:
```suggestion
From figure 1, the backfill task is firstly split into serial sub tasks, each sub tasks include one batch size record. Then, multi-workers are started up to execute backfill sub tasks parallelly.
```

File: docs/design/2022-06-06-Adding-Index-Acceleration.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+   ...
+ 
+   // Collect duplicate data from TiKV.
+   rpc DuplicateDetect(DuplicateDetectRequest) returns (stream DuplicateDetectResponse) {}
+}
+```
+Duplicate key will be reported when two unique index records with the same key but different value are encountered. In this case, we could report the error and cancel this DDL job.
+
+### Checkpoints
+In current implementation, the progress of creating an index is persisted in reorgCtx, it is updated when each batch of the index is backfilled. For new backfiller in this proposal, we could mark the progress in a similar way when a SST file generated from a batch of index is ingested successfully.
```
Comment (zimulala) [inline docs/design/2022-06-06-Adding-Index-Acceleration.md:76]:
```suggestion
In the current implementation, the progress of creating an index is persisted in reorgCtx, it is updated when each batch of the index is backfilled. For the new backfiller in this proposal, we could mark the progress in a similar way when an SST file generated from a batch of indexes is ingested successfully.
```

File: docs/design/2022-06-06-Adding-Index-Acceleration.md [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+- Each TiDB instance should be able to check the inactive SST files and clean them, in case ownership of the job execution changes because of TiDB instance crashed.
+
+## Limitations and future works
+Compared with the current one, the CPU and memory consumption of new backfiller is not controlled smoothly when doing the ‘sort and ingest’.
+
+## Compatibility
+As an optimization, this proposal does not lead to any semantic change so theoretically there will be no compatibility issue.
+
+### Enable/Disable
+1. Use tidb_ddl_enable_fast_reorg to enable and disable new solution, default value is off.
```
Comment (bb7133) [inline docs/design/2022-06-06-Adding-Index-Acceleration.md:92]: IMHO it's better to reformat it in the following way:

1. Use `tidb_ddl_enable_fast_reorg` to enable and disable the new solution, the default value is off. Example:

```
TiDB> set global tidb_ddl_enable_fast_reorg = on/off;
```

Same as item 2 and 3.

File: docs/design/2022-06-06-Adding-Index-Acceleration.md [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+    Example:```TiDB> set global tidb_ddl_enable_fast_reorg = on/off;```
+
+2. Use `tidb_ddl_disk_quota` to set a threshold of local disk max usage by new backfiller, the range of this variable is from 100 GiB to 1 PiB, default value is 100 GiB.
+   
+    Example:```TiDB> set global tidb_ddl_disk_quota = 107374182400;```
+
+3. Use one config parameter `fast-reorg-local-path` to specify the sort path for lightning when it used in add index reorg stage. The default value is /tmp/tidb, please change it to your own disk path for the new backfiller could work properly.
+
+    Example:```fast-reorg-local-path = “/tmp/tidb”```
```
Comment (bb7133) [inline docs/design/2022-06-06-Adding-Index-Acceleration.md:102]:
```suggestion
    Example:```fast-reorg-local-path = "/tmp/tidb"```
```

**Review by bb7133 — APPROVED** [review]
LGTM

**Pull request merged**